packages/zoe/src/zoe.chainmail

```diff
@@ -42,60 +47,52 @@ interface ZoeService {
+   * by Zoe to enforce offer safety, and `exit` is used to specify
    * the extent of payout liveness that Zoe can guarantee.
-   * `payoutRules` is a list of objects, each with a `kind` property 
-   * (such as 'offerAtMost') and an amount property. The objects in the 
-   * `payoutRules` must be in the same order
-   * as the issuers associated with a smart contract. That is, the
-   * amount in index 0 of `payoutRules` should be an amount for the issuer
-   * in index 0 of the issuers array. `payments` is an array of the
-   * actual payments to be escrowed, following the rules in the
-   * payoutRules. If the payoutRules kind is 'offerAtMost',
-   * then a payment matching the amount is expected. If the payoutRules
-   * kind is 'wantAtLeast' then the payment will be
-   * ignored and should be `undefined`.
+   * `want` and `offer` are objects with keywords as keys and amounts
+   * as values. `payments` is a record with keywords as keys,
+   * and the values are the actual payments to be escrowed. A payment
+   * is expected for every rule under `offer`.
    */
-  redeem (invite :Invite, offerRules :OfferRules, payments :List(Payment)) -> (SeatAndPayout);
+  redeem (invite :Invite, offerRules :OfferRules, payments :PaymentKeywordRecord)) -> (SeatAndPayout);
 }
 
 /**
  * This is returned by a call to `redeem` on Zoe.
  * A seat is an arbitrary object whose methods allow the user to take
- * certain actions in a contract. The payout is a promise that resolves to an array
- * of promises for payments. Note that while the payout promise
- * resolves when an offer is completed, the promises for each payment
- * resolves after the remote issuer successfully withdraws the payment.
+ * certain actions in a contract. The payout is a promise that resolves
+ * to an object which has keywords as keys and promises for payments
+ * as values. Note that while the payout promise resolves when an offer
+ * is completed, the promise for each payment resolves after the remote
+ * issuer successfully withdraws the payment.
  */
 struct SeatAndPayout ( ) {
   seat :Object;
   payout :List(Payment);
 }
 
 struct OfferRules ( ) {
-  payoutRules :List(PayoutRule);
-  exitRule :ExitRule;
+  want :AmountKeywordRecord;
+  offer :AmountKeywordRecord;
+  exit :ExitRule;
 }
 
 /**
- * payoutRules are an array of PayoutRule. The possible
- * kinds are 'offerAtMost' and 'wantAtLeast'.
+ * The keys are keywords, and the values are amounts. For example:
+ * { Asset: amountMath.make(5), Price: amountMath.make(9) }
  */
-struct PayoutRule ( ) {
-  kind :PayoutRuleKind;
-  amount :Amount;
+struct AmountKeywordRecord ( ) {
 }
 
 /**
- * The possible kinds are 'waived', 'onDemand', and 'afterDeadline'.
- * `timer` and `deadline` only are used for the `afterDeadline` kind.
+ * The possible keys are 'waived', 'onDemand', and 'afterDeadline'.
+ * `timer` and `deadline` only are used for the `afterDeadline` key.
```

Comment:
I don't think we have a good way to document optional fields in chainmail, but I'm not sure typescript will be better. I think it would be good to show (maybe in this comment) that the possibilities are 

```
struct ExitRule () { waived: '' }
struct ExitRule () { onDemand: '' }
struct ExitRule () { afterDeadline: '', timer :Timer, deadline :Number }
```

What are the expected values for `waived`, `onDemand`, and `afterDeadline`?

Reply:
I like this idea a lot - I'll do that.